velox/docs/develop/memory.rst

```diff
@@ -0,0 +1,890 @@
+   The query pool uses the default implementation which sorts its child task
+   pools based on the reclaimable bytes (*MemoryPool::reclaimableBytes*), and
+   reclaim from the task with the most reclaimable bytes until reaches the
+   reclaim target
+
+#. The query pool invokes the reclaim method of the task pool which in turn
+   calls into the associated task reclaimer (*Task::MemoryReclaimer*). The
+   latter first pauses the task execution (*Task::requestPause*), and then
+   sorts its child node pools based on the reclaimable bytes and reclaims
+   memory from the node pools with the most reclaimable bytes. After reaching
+   the reclaim target or having reclaimed from all the node pools, task
+   reclaimer resumes the task execution (*Task::resume*)
+
+#. The task pool invokes the reclaim method of the node pool which reclaim
+   memory from its child operator pool with the most reclaimable bytes
+
+#. The node pool eventually calls the operator pool to do the actual memory
+   reclamation (*Operator::MemoryReclaimer*). Currently we support memory
+   reclamation through disk spilling and table writer flush. *Operator::reclaim*
+   is added to support memory reclamation with the default implementation does
+   nothing. Only spillable operators override that method: *OrderBy*, *HashBuild*,
+   *HashAggregation*, *RowNumber*, *TopNRowNumber*, *Window* and *TableWriter*.
+   As for now, we simply spill everything from the spillable operator’s row
+   container to free up memory. After we add memory compaction support for row
+   containers, we could leverage fine-grained disk spilling features in Velox
+   to only spill and free the required amounts of memory.
+
+Note memory arbitrator can’t reclaim from a spillable operator if it has
+triggered memory arbitration in the middle of data processing even after it
+has stopped its query task execution. To prevent this, we added
+*Operator::nonReclaimableSection_* to indicate if an operator is under a
+non-reclaimable section or not, and the memory arbitrator can’t reclaim memory
+from an operator which is under a non-reclaimable section. The driver execution
+framework sets a running operator in the non-reclaimable section by default.
+The spillable operator chooses to clear the non-reclaimable section at specific
+call sites such as the memory reservation (*MemoryPool::maybeReserve*) before the
+actual data processing to allow the memory arbitrator to reclaim memory.
+
+Memory Allocator
+----------------
+The memory allocator manages the physical memory allocations for both query
+memory allocated through memory pool and cache memory allocated directly from
+file cache. The memory allocator ensures the total allocated memory is always
+within the system memory limit. *MemoryAllocator* defines the memory allocator
+interface. We have two allocator implementations: *MallocAllocator* delegates
+all the memory allocations to std::malloc which is simple and reliable. We
```

Comment:
nit: might be helpful to clarify that it is possible to configure the application to use jemalloc as well.

Reply:
Added "Users can choose the
allocator for their application by setting *MemoryManagerOptions::useMmapAllocator*
(see for `memory system setup section <#memory-system-setup>`_ for example)" at end of this paragraph